parsl/providers/azure/azure.py

```diff
@@ -0,0 +1,478 @@
+import json
+import logging
+import os
+import time
+from string import Template
+
+from parsl.dataflow.error import ConfigurationError
+from parsl.providers.azure.template import template_string
+from parsl.providers.provider_base import ExecutionProvider
+from parsl.providers.error import OptionalModuleMissing
+from parsl.utils import RepresentationMixin
+from parsl.launchers import SingleNodeLauncher
+
+logger = logging.getLogger(__name__)
+
+try:
+    from azure.common.credentials import ServicePrincipalCredentials
+    from azure.mgmt.resource import ResourceManagementClient
+    from azure.mgmt.network import NetworkManagementClient
+    from azure.mgmt.compute import ComputeManagementClient
+    from azure.mgmt.compute.models import DiskCreateOption
+    from msrestazure.azure_exceptions import CloudError
+
+    _api_enabled = True
+
+except ImportError:
+    _api_enabled = False
+else:
+    _api_enabled = True
+
+translate_table = {
+    'VM pending': 'PENDING',
+    'VM running': 'RUNNING',
+    'VM deallocated': 'COMPLETED',
+    'VM stopping': 'COMPLETED',  # We shouldn't really see this state
+    'VM stopped': 'COMPLETED',  # We shouldn't really see this state
+}
+
+
+class AzureProvider(ExecutionProvider, RepresentationMixin):
+    """
+    A Provider for using Microsoft Azure Resources
+
+    One of 2 methods are required to authenticate: keyfile, or environment
+    variables. If  keyfile is not set, the following environment
+    variables must be set: `AZURE_CLIENT_ID` (the access key for
+    your azure account),
+    `AZURE_CLIENT_SECRET` (the secret key for your azure account), the
+    `AZURE_TENANT_ID` (the session key for your azure account), and
+    AZURE_SUBSCRIPTION_ID.
+
+    Parameters
+    ----------
+    vm_reference : dict
+        Dictionary describing the parameters of the Azure VM.
+
+        Required structure:
+        {
+          'publisher': VM OS publisher
```

Comment:
I found it quite awkward to find a userful publisher/offer/sku/version combo - it involved a lot of messing round in the Azure Marketplace. I don't know if there is a better way. But make some note about where to get these values from.